### ISSUE TITLE
CSS Multi-column Layout

### ISSUE THREAD

**Issue by SimonSapin** [Kozea/WeasyPrint#60]
http://dev.w3.org/csswg/css3-multicol/

Column balancing is tricky, the rest is relatively straightforward. We should start with multicol without `column-fill` (balancing) or `column-span` (which also requires balancing).

**Comment by thinkwelltwd** [comment]
I'm not able to develop multicol support, but would be interested in discussion financial support for this feature.

**Comment by liZe** [comment]
> I'm not able to develop multicol support, but would be interested in discussion financial support for this feature. Who would be the right person to talk to about that?

Sorry for being so long. If anyone is interested in financial support, you can contact me (guillaume.ayoub[at]kozea.fr).

**Comment by StephanMeijer** [comment]
I'd really like this functionality. +1

**Comment by liZe** [comment]
I've added a really simple multi-column layout support.

What's implemented:
- the pseudo-algorithm defining how many columns must be used,
- the replacement of normal layout with multi-column layout when needed,
- the layout of the columns,
- a stupid balancing algorithm,
- multi-column blocks cut between multiple pages.

What can be probably easily added:
- paddings, borders, margins,
- management of column-gap and column-rule-* properties.

What seems to be a bit harder (and less interesting):
- a real shrink-to-fit algorithm (the spec says that "CSS does not define the exact algorithm", funny as usual),
- a clean way to add column boxes (it's currently added with a huge hammer),
- check that everything goes well with the various values of the "position" and "display" properties on the container (for ex. check that the columns are applied to "non-replaced block-level elements (except table elements), table cells, and inline-block elements"),
- check that everything goes well with the various values of the "position" and "float" on the children,
- check that everything goes well when a block is split between multiple pages,
- forced column breaks and page breaks,
- column spans.

What seems to be really hard:
- a better balancing algorithm (contrary to what I was thinking, without this the feature is actually pretty useless),
- break-inside, break-before and break-after in columns (other than forced breaks),
- orphans and widows in columns.

I'm testing this branch with [WeasySuite](http://test.weasyprint.org), the tests from the W3 suite seem to be really good.

I'll implement the easy part soon and add some tests (and renderings) in this issue.

**Comment by liZe** [comment]
I've added the easy part, everything is in the `columns` branch.

An ugly sample, with colors, margins, borders, paddings, hyphenation, gap, rule and justification:
```html
<html lang="en">
<style>
body { -weasy-hyphens: auto; font-size: 10pt }
div { columns: 3; padding: 1em; background: red; background-clip: content-box; border: 3px solid; text-align: justify; column-gap: 3em; column-rule: 1px dashed blue }
p { background: yellow; padding: 0.5em }
</style>
<div>
<p>A multi-column element (or multicol element for short) is an element whose ‘column-width’ or ‘column-count’ property is not ‘auto’ and therefore acts as a container for multi-column layout.
<p>In the traditional CSS box model, the content of an element is flowed into the content box of the corresponding element. Multi-column layout introduces a new type of container between the content box and the content, namely the column box (or column for short). The content of a multicol element is flowed into its column boxes.
<p>Column boxes in a multi-column element are arranged into rows. Like table cells, the column boxes in a row are ordered in the inline direction of the multicol element. The column width is the length of the column box in the inline direction. The column height is the length of the column box in the block direction. All column boxes in a row have the same column width, and all column boxes in a row have the same column height. Within each row in the multi-column element, adjacent column boxes are separated by a column gap, which may contain a column rule. All column gaps in the same row are equal. All column rules in the same row are also equal, if they appear; column rules only appear between columns that both have content.
<p>In the simplest case a multicol element will contain only one row of columns, and the height of each column will be equivalent to the used height of the multi-column element's content box. If the multi-column element is paginated, the height of each row is constrained by the page and the content continues in a new row of column boxes on the next page; a column box never splits across pages. The same effect occurs when a spanning element divides the multi-column element: the columns before the spanning element are balanced and shortened to fit their content. Content after the spanning element then flows into a new row of column boxes.
<p>It is not possible to set properties/values on column boxes. For example, the background of a certain column box cannot be set and a column box has no concept of padding, margin or borders.
```

![test](https://cloud.githubusercontent.com/assets/952243/17653757/47c7e418-629a-11e6-9685-ba8f221c4f8d.png)

There's a lot of things to do, help and reviews are welcome!

**Comment by liZe** [comment]
You can take a look at [the diff](https://github.com/Kozea/WeasyPrint/compare/columns).

**Comment by liZe** [comment]
It's now quite clean and ready to be tested. Some of the tests of the W3C test suite pass, and many simple real-life results could be really good.

**Comment by mb21** [comment]
wow, cool :) I'm looking forward to take a better look at this in the next couple of weeks!

**Comment by liZe** [comment]
I've added the W3C test suite to the next version of WeasySuite, the results are pretty good for a feature that has just been added (42% tests pass). Solving many stupid corner cases (mostly about collapsing margins and absolute positioning) will easily make this number reach 50~60%.

After playing with the implementation, I think that column-breaking is much easier than page-breaking, and that this feature could be added in a couple of hours. Orphans and widows are painful, but possible. Column spans are even more painful.

A real balancing algorithm would probably take days to imagine and to code, with a small benefit. That's not my priority for the next months.

And about shrink-to-fit, well… I'm sure that David Baron would be delighted to solve this easy problem for us :wink:.

If anyone is interested in testing and/or adding new features and/or fixing bugs, I'll be happy to help!
